cmd/assetsvc/postgresql_utils.go

```diff
@@ -235,34 +213,77 @@ func (m *postgresAssetManager) getChartFilesWithFallback(namespace, filesID stri
 	return chartFiles, nil
 }
 
-func containsVersionAndAppVersion(chartVersions []models.ChartVersion, version, appVersion string) (models.ChartVersion, bool) {
-	for _, ch := range chartVersions {
-		if ch.Version == version && ch.AppVersion == appVersion {
-			return ch, true
-		}
+func (m *postgresAssetManager) getPaginatedChartListWithFilters(cq chartQuery, pageNumber, pageSize int) ([]*models.Chart, int, error) {
+	whereQuery, whereQueryParams := m._generateWhereClause(cq)
+	charts, numPages, err := m._getPaginatedChartList(whereQuery, whereQueryParams, pageNumber, pageSize)
+	if err != nil {
+		return []*models.Chart{}, 0, err
 	}
-	return models.ChartVersion{}, false
+	return charts, numPages, nil
 }
 
-func (m *postgresAssetManager) getChartsWithFilters(namespace, chartName, version, appVersion string) ([]*models.Chart, error) {
-	clauses := []string{"info ->> 'name' = $1"}
-	queryParams := []interface{}{chartName, namespace}
-	if namespace != dbutils.AllNamespaces {
-		queryParams = append(queryParams, m.GetKubeappsNamespace())
-		clauses = append(clauses, "(repo_namespace = $2 OR repo_namespace = $3)")
-	} else {
-		clauses = append(clauses, "repo_namespace = $2")
+func (m *postgresAssetManager) _generateWhereClause(cq chartQuery) (string, []interface{}) {
+	whereClauses := []string{}
+	whereQueryParams := []interface{}{}
+	whereQuery := ""
+
+	if cq.namespace != dbutils.AllNamespaces {
+		whereQueryParams = append(whereQueryParams, cq.namespace, m.GetKubeappsNamespace())
+		whereClauses = append(whereClauses, fmt.Sprintf(
+			"(repo_namespace = $%d OR repo_namespace = $%d)", len(whereQueryParams)-1, len(whereQueryParams),
+		))
 	}
-	dbQuery := fmt.Sprintf("SELECT info FROM %s WHERE %s ORDER BY info ->> 'ID' ASC", dbutils.ChartTable, strings.Join(clauses, " AND "))
-	charts, err := m.QueryAllCharts(dbQuery, queryParams...)
-	if err != nil {
-		return nil, err
+	if cq.chartName != "" {
+		whereQueryParams = append(whereQueryParams, cq.chartName)
+		whereClauses = append(whereClauses, fmt.Sprintf(
+			"(info->>'name' = $%d)", len(whereQueryParams),
+		))
 	}
-	result := []*models.Chart{}
-	for _, c := range charts {
-		if _, found := containsVersionAndAppVersion(c.ChartVersions, version, appVersion); found {
-			result = append(result, c)
+	if cq.version != "" && cq.appVersion != "" {
+		whereClauses = append(whereClauses, fmt.Sprintf(
+			"(info->'chartVersions' @> '[{\"version\":\"%s\",\"app_version\":\"%s\"}]'::jsonb)", cq.version, cq.appVersion,
```

Comment:
this could be a SQL injection problem (we are not escaping version nor appVersion from the request, right?)

Reply:
Absolutely agree with you. I tried to add $-parameters, but Postgresql does not allow to parameterize inside of a jsonb literal and I ended by simply doing so; totally my bad! Thanks for pointing this out.

Changing it to `(info->'chartVersions' @> $1::jsonb)`, where `$1` would be `fmt.Sprintf('[{"version":"%s","app_version":"%s"}]', cq.version, cq.appVersion)`
Therefore, the injection cannot break out of scope of $, I guess.